No videos play if opened in full screen on first play

pvollan opened the pull request:
#### 62b9ab75af70fc1cb9ada689c9f3f482a0274e30
<pre>
No videos play if opened in full screen on first play
<a href="https://bugs.webkit.org/show_bug.cgi?id=255048">https://bugs.webkit.org/show_bug.cgi?id=255048</a>
rdar://107540521

Reviewed by Eric Carlson and Jean-Yves Avenard.

When CA layers are not remotely hosted in both the GPU process and the WebContent process, it is not
guaranteed that the remote context ID will be set in the HTML media element when entering fullscreen.
This patch works around that by delaying setting up the fullscreen until the video dimensions are set.
At this point, the remote context ID should be available.

* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.h:
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::VideoFullscreenManager::enterVideoFullscreenForVideoElement):
(WebKit::VideoFullscreenManager::videoDimensionsChanged):

Canonical link: <a href="https://commits.webkit.org/262654@main">https://commits.webkit.org/262654@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c6e1cb814ede546124a0c95d861d4a26d52ddd1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2134 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2166 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3059 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2188 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2246 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2211 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1940 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2153 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1909 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1914 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2911 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1894 "1 api test failed or timed out") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1895 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1796 "143 failures") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1968 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1943 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3078 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1960 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1757 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1887 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1905 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/538 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2072 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.h:
```diff
@@ -190,6 +192,7 @@ class VideoFullscreenManager : public RefCounted<VideoFullscreenManager>, privat
     HashMap<PlaybackSessionContextIdentifier, int> m_clientCounts;
     WeakPtr<WebCore::HTMLVideoElement, WebCore::WeakPtrImplWithEventTargetData> m_videoElementInPictureInPicture;
     bool m_currentlyInFullscreen { false };
+    WTF::Function<void(LayerHostingContextID)> m_setupFullscreen;
```
Can this be a CompletionHandler?

pvollan replied:
> Can this be a CompletionHandler?

Yes, I think it could. I have kept is as Function, though, since I think both make sense in this context.

eric-carlson commented on Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.h:
```diff
@@ -190,6 +192,7 @@ class VideoFullscreenManager : public RefCounted<VideoFullscreenManager>, privat
     HashMap<PlaybackSessionContextIdentifier, int> m_clientCounts;
     WeakPtr<WebCore::HTMLVideoElement, WebCore::WeakPtrImplWithEventTargetData> m_videoElementInPictureInPicture;
     bool m_currentlyInFullscreen { false };
+    WTF::Function<void(LayerHostingContextID)> m_setupFullscreen;
```
Nit: I'm not wild about this name, maybe `m_setupFullscreenHandler`?

pvollan replied:
Fixed in latest patch.

Thanks for reviewing!

jyavenard commented on Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
```diff
@@ -454,6 +464,16 @@ static FloatRect inlineVideoFrame(HTMLVideoElement& element)
+
+    if (m_setupFullscreen) {
+        auto [model, interface] = ensureModelAndInterface(contextId);
+        RefPtr<HTMLVideoElement> videoElement = model->videoElement();
```
auto videoElement

need null check

pvollan replied:
Good catch! Fixed in latest patch.

Thanks for reviewing!

jyavenard reviewed (approved):
typo in the commit log:
s/guranteed/guaranteed 

some comments

pvollan committed:
No videos play if opened in full screen on first play
https://bugs.webkit.org/show_bug.cgi?id=255048
rdar://107540521

Reviewed by Eric Carlson and Jean-Yves Avenard.

When CA layers are not remotely hosted in both the GPU process and the WebContent process, it is not
guaranteed that the remote context ID will be set in the HTML media element when entering fullscreen.
This patch works around that by delaying setting up the fullscreen until the video dimensions are set.
At this point, the remote context ID should be available.

* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.h:
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::VideoFullscreenManager::enterVideoFullscreenForVideoElement):
(WebKit::VideoFullscreenManager::videoDimensionsChanged):

Canonical link: https://commits.webkit.org/262654@main
